### ISSUE TITLE
Adding an Indexer#load_config method?

### ISSUE THREAD

**Issue by jrochkind** [traject/traject#99]
Currently, if working with a Traject::Indexer, you can load a config file into it located at `config_path` by:

     indexer.instance_eval( File.open(config_path).read, config_path)

(The second arg makes sure stack traces from the code in the config file have the real config file path in them). 

It's possible this is good enough, if it were documented clearly. It's also possible this is just confusing enough that a wrapper method would be convenient. We start with, on Traject::Indexer, something like:

~~~ruby
def load_config(file_path)
    self.instance_eval(File.open(file_path).read, file_path)
end
~~~

I started down this path, but the tricky part is exceptions. There are two main categories of exceptional condition that could happen: 1) No file is found there, or the file can not be read, 2) There is a syntax error or other exception raised when evaluating the config file. 

I am not sure if this method should just allow the 'natural' exceptions to be raised, and caught by the caller if desired. Or if this method should rescue the exceptions, and wrap them in a custom Traject::Indexer::Exception of some kind, for predictable exception classes.  Nearly any exception at all could hypothetically be raised in the second category of circumstance (error of some kind when evaluating the config file). 

I can see an argument both ways, I'm not sure what is best. Since nearly any exception possible could be raised, we might need to rescue `Exception`, not just `StandardError`.  (SyntaxError for instance is not a StandardError).  Or maybe we could rescue StandardError and ScriptError. It gets confusing. 

Here's what CommandLine is doing, it does rescue Exception generally, even though that's kind of bad form, I just punted on it, and figured it was okay for CommandLine, as the code is only operative at beginning of program execution when loading config files. https://github.com/traject/traject/blob/master/lib/traject/command_line.rb#L173-L194

Any thoughts? @billdueber ? Anyone else?

**Comment by billdueber** [comment]
Boy, I'd be tempted to just let missing/unreadable file exceptions bubble through -- there's no meaningful rescue, and the standard "file not found" error is at least as meaningful as anything else. 

For errors in `instance_eval`, though, we'll probably want to catch it and then raise a more meaningful error, with the name of the file being sucked in and so on. Make sense?

**Comment by jrochkind** [comment]
Hm. You've got to catch almost _everything_ to catch errors in instance_eval though, probably `rescue StandardError, SyntaxError => e`.  

What is it that would make a custom error 'more meaningful' than whatever was originally raised? The name of the file being sucked in will already be in the stack trace, along with the line number where it raised, in the original 'natural' error.  

Could trying to raise a custom error actually end up _hiding_ or suppressing information, and being less meaningful? The original error of course, is the _real_ error.  If we did raise a custom consistent error, would we want to set the stack trace to the same as the original exception (this actually is possible in ruby, little known fact).  So the dev could see where it was really coming from, the purpose of the stack trace? But then wouldn't you want the class and message to be maintained too, precisely for the information they carry about what went wrong?  

I'm partially playing devil's advocate here, neither choice seems obviously right to me, trying to work through it.

**Comment by jrochkind** [comment]
I _might_ have an idea to get the best of both worlds, I'll try it out. But curious if you have any opinion or feedback on what makes a rescue-and-raise-own-exception more meaningful and/or convenient than just letting the natural exceptions raise. 

(And if we in fact do need this convenience api instead of just the raw instance_eval?)

**Comment by billdueber** [comment]
Addressed in #100; closing.
